### ISSUE TITLE
Connect to Daly BMS via TTL/UART

### ISSUE THREAD

**Issue by sanitariu** [fl4p/batmon-ha#225]
Why not allow direct connection via TTL to your PI3/4 device ?
It is possible and you can read data directly from Daly BMS. No need for wireless.

**Comment by Off-Grid-Garage** [comment]
This is not part of this project.

Some users don't have the Pi near their BMS and in my case, I have 12BMS plus another 10 in mobile 12V batteries which I need to monitor. Cabling impossible.

**Comment by gnuorder** [comment]
OGG, I just watched your video. Most of us don't have 22 BMS to monitor and as you found, Bluetooth just isn't reliable enough. I'm still looking for a good solution that can look at several BMS through RS485/TTL/CAN. for more reliable monitoring that doesn't require a MCU for each BMS. I currently run the standalone program, HybridInverterMon, which can monitor 1 JKBMS over RS485 (Im working on adding additional JKBMS but I'm not familiar with C# or DOTNET) and not any other brand BMS. If this project can be expanded to include RS485 communication it sounds good for even up to 10 BMS. I also don't know why peterboards require 1 board per BMS because I thought CAN also supports multi devices on a bus. You of course have your batteries all over the garage for testing and what not so cabling would be a mess but for someone like me who only plan on a few batteries in the same space, this would be ideal.

**Comment by fl4p** [comment]
It is on my todo, since most BMS just have the BT dongle sit on the UART, we just need to change the transport layer, the protocol should be the same

also see https://github.com/fl4p/batmon-ha/issues/22

**Comment by Off-Grid-Garage** [comment]
@typxxi coming soon on my channel...

**Comment by fl4p** [comment]
@typxxi I've added experimental wrapper for wired serial communication. https://github.com/fl4p/batmon-ha/commit/7471db219b26ab4897ab459473c0f0c7d8b28236
You'll need to set device `address: serial` and `adapter: <serial port>`:
```yaml
address: serial
type: jk
adapter: /dev/cu.usbmodem1234
 ```

Notice that I didn't test it and it just writes all the BLE commands to the serial port, and passes all received data to the BLE notification callbacks. As everything is checksummed, there is low risk that anything go wrong. Still, don't try this in a live setup and always review BMS settings afterwards.

**Comment by typxxi** [comment]
Thanks a lot

> You'll need to set device `address: serial` and `adapter: <serial port>`:

So I will use an UART TTL - USB Adapter and connect that UART like that shown [here ](https://github.com/syssi/esphome-jk-bms/discussions/715#discussioncomment-12649276) in this solution ?

![Image](https://github.com/user-attachments/assets/9d86dc6b-d33f-480c-8bcc-2021a753ebee)

I simply use a USB FTDI Adapter  - but can I use the RS485-2 ports for the internal bms communication on the right as shown here on the righ the 2 ports ?

![Image](https://github.com/user-attachments/assets/cfa7e63e-903a-4301-a78e-3faa8b55ef8d)

**Comment by fl4p** [comment]
Yeah, any USB-UART adapter will work.

> I simply use a USB FTDI Adapter - but can I use the RS485-2 ports for the internal bms communication on the right as shown here on the righ the 2 ports ?
I don't understand this.

**Comment by typxxi** [comment]
If you look at the picture of the front panel then there are 2 RS485-2 ports next to each others which are usually used to parallel 3 JK BMS and batteries this way. YOu will have a master and all others are slaves, hence you have to deal only with 1 BMS instead of 3 BMS settings. Master will send the settings to the slaves. 

I am not sure if I could use these for this. I have not started it cause I am still abroad and might return next week(end).

**Comment by fl4p** [comment]
@sanitariu The Daly UART port appears to be lacking isolation so its a good practice to add signal isolation. It is probably even necessary since the BMS switches Bat(-) which is usually GND. The serial link GND might short-cut the BMS switch, which should release magic smoke.

I found cheap isolated USB-TTL adapters, look for "ft232 isolated".

And this article: https://pedrogomez.co/blog/2021/04/21/replacement-optoisolated-usb-uart-daly-smart-bms


MIKROE also sells DIGI ISOLATOR CLICK board 

![Image](https://github.com/user-attachments/assets/c17ef558-0cc3-47eb-8fcb-21339ee29828)

**Comment by fl4p** [comment]
it has been implemented:
```yaml
- address: serial
  adapter: /dev/ttyUSB0   # serial port path; required when address=serial
  type: jk_uart
  alias: battery1
```

**Comment by typxxi** [comment]
Thanks a lot for taking care as it is still missing and I had reliability issues from time to time (even though 20 cm distance between external Pi4 antenna in front of the battery on the right corner).

**Comment by fl4p** [comment]
did you try it and it worked?

**Comment by typxxi** [comment]
Sorry, I have not and just wanted to say thank you here - as I had asked on another topic the "modbus network chain", where you had explained, that ti would not work. I have asked another question about the proper RJ11 RS232 ftdi adapter and if that would work with CAN BUS connecting to the growatt inverter and RS485 -2 connecting to my Windows Laptop.

Once you have replied I will order and install next week I guess when the piece will have arrive. I will need 2 or 3 as I have already 2 JK inverter BMS batteries with Growatt SPH. Bluetooth had a horrible time last year but most recently it worked a lot better - out of nowhere. But proper BMS data is too important than struggling with all the breaks I had faced before via 3 or 4 BT adapters incl. external Antenna and literally next to the right terminal of the EEL Case.

Thanks

**Comment by fl4p** [comment]
If you ever want to give Bluetooth another try: use the new 'bluek' or 'bumble' stack. bumble uses exclusive adapter access and talks directly to the hardware skipping anything in between (bleak, dbus, bluez, bluetoothd). bluek skips bleak and dbus.
If issues persists its probably your BMS BLE adapter, I've seen nasty things.

The wired approach is def the safest. I was recently struggiling with BLE<->WIFI interference.

**Comment by typxxi** [comment]
> If you ever want to give Bluetooth another try: use the new 'bluek' or 'bumble' stack. bumble uses exclusive adapter access and talks directly to the hardware skipping anything in between (bleak, dbus, bluez, bluetoothd). bluek skips bleak and dbus. If issues persists its probably your BMS BLE adapter, I've seen nasty things.
> 
> The wired approach is def the safest. I was recently struggiling with BLE<->WIFI interference.

### **I. WIRED TRIALS**

Finally I have tried the "wired connection" yesterday really hard.
I struggled a lot even with AI support as it felt that the old BT connection seem to be working hidden. Weird description but it felt to me like this hidden bt was acting in the background. Also configuration changes (via UI or yaml) did not always save everything. I changed the intervalls from 3 - 3 - 20 to 4 - 4 - 30 seconds (sample, publish, expire) but after pressing save , clicking on the tab info and then clicking on the configuration tab only the first 4 had been saved to 4 - 3- 20 instead of 4-4-30.
I tried it multiple times and the issue was still there, but slightly different. Sometimes 2 values had been changed or even all. 
Yes, I have uninstalled the add on incl. deleting all data, rebooted HA system and readded the whole addon , but still from time to time all changes had not been saved. This is also not depending on the yaml or UI mode, cause it also happened in UI mode that not everything was there after saving. 

Those are my observations during the new wired setup trials till I was simply too tired, gave up and tried to go back to my previous config from yesterday morning. 

```
devices:
  - address: C8:47:80:10:08:A1
    type: jk
    alias: battery1
    current_calibration: 1
concurrent_sampling: false
keep_alive: true
verbose_log: false
sample_period: 4
publish_period: 4
invert_current: true
watchdog: true
expire_values_after: 20
bt_power_cycle: true
ble_stack: bleak
bluetooth_proxies: []

```
That did not work out that way as I had hoped. 
I had to add more lines ... even though there was no cable connection anymore and the USB bluetooth dongle had been there connected all the time in the past 15 or 18 months.

I went to sleep at 3 am in the hope everything would be fine, but during sunday I found several occasions a full stop of the addon which I might have not recogniced before cause watchdog might have been active and restarted automatically -  I guess 

### **II. NEW GOALS: 3 batteries will require a cable connection to HA**

I rarely saw these full stops or struggling of Batmon when BMS data like battery temperature were missing in the dashboard from time to time, where I usually did not care that much as they came back at leeast after a HA restart. I did not investigate that till I started the preparation of 2 additional batteries, each 16 kWh and all JK BMS - EEL V4 batteries. 
At that moment in time I was aware that I have to rebuild a lot as I have to replace my Growatt grid tied inverter with 2 SPH hybrid ones that have to communicate and also each with their batteries. 

At that point I did no longer wanna  rely on bluetooth risks with 3 inverters and 3 batteries in the same 3x 3 m room inclunding the HA Pi 4 server (connected via LAN not WLAN).  

### **III. A LOG FROM THE "PRE BUMBLE" CONFIGURATION**

Now i have at least a log file of sunday evening showing 1 hour from 18:25 to 19:25 
which I have attached. Should be a pure BT connection via an external USB BT dongle while the internal one had been disabled like in the months before. 

[batmon logs 18.25 - 19.26.txt](https://github.com/user-attachments/files/28688070/batmon.logs.18.25.-.19.26.txt)

### **IV. THE "BUMBLE" CONFIGURATION TRIALS**

When I got back to your comment above today I finaly read about your explanations regarding bumble and I have now activated bumble - everything els stayed the same, but this WAS NOT PART OF THE SETUP WHEN THE LOG WAS RECORDED.

```
devices:
  - address: C8:47:80:10:08:A1
    type: jk
    alias: battery1
    current_calibration: 1
    adapter: hci0
concurrent_sampling: false
keep_alive: true
verbose_log: true
sample_period: 3
publish_period: 3
invert_current: true
watchdog: true
expire_values_after: 20
bt_power_cycle: true
ble_stack: bumble
bluetooth_proxies: []
```

and here is my bluetooth integration setup too which made me wonder as the batmon log still shows these lines for 2 adapters not just 1 ... even though the hci0 is the right one that is also showing the BMS HBB_

23:30:46 INFO [bt] BT Discovery (5 seconds, adapter=hci0):
23:30:46 INFO [bt] BT Discovery (5 seconds, adapter=hci1):
23:30:51 INFO [bt] hci1               addr                       name rssi
23:30:51 INFO [bt] hci1  6C:79:D8:F5:A3:2F                       None  -31
23:30:51 INFO [bt] hci1  78:11:DC:AA:6A:64                       None  -50
23:30:51 INFO [bt] hci1  C8:47:80:10:08:A1                       None  -54
23:30:51 INFO [bt] hci0               addr                       name rssi
23:30:51 INFO [bt] hci0  06:87:9E:5A:D7:7C                       None  -92
23:30:51 INFO [bt] hci0  4C:65:A8:D2:2B:87                   MJ_HT_V1  -84
23:30:51 INFO [bt] hci0  6C:79:D8:F5:A3:2F                       None  -44
23:30:51 INFO [bt] hci0  78:11:DC:AA:6A:64                     mi_mtk  -68
23:30:51 INFO [bt] hci0  8C:F6:81:31:15:8B                   WB251327 -100
23:30:51 INFO [bt] hci0  BC:35:1E:B4:90:80                       None  -86
23:30:51 INFO [bt] hci0  C8:47:80:10:08:A1           HBB_PB2A16S20-01  -50
23:30:51 INFO [bt] hci0  FC:03:9F:25:34:A3                       None  -90


<img width="1168" height="605" alt="Image" src="https://github.com/user-attachments/assets/ef07c7c9-9efd-41b0-a030-56f1d484b29a" />

Here is the log file.
[batmon logs 'bumble' - 23.30.txt](https://github.com/user-attachments/files/28688322/batmon.logs.bumble.-.23.30.txt)

### **V. THE NEW "BUMBLE" CONFIGURATION TRIALS with 4-4-20 seconds**

And here is the latest log from the running "ble_stack: bumble" but with reconfiguration from 3 to 4 to 5 after still facing issue - this time even within minutes somehow. But as a reminder : i did not reboot the whole system every single time, just 2 times today.

[batmon logs 'bumble' 4-4-20 seconds - 23.38.txt](https://github.com/user-attachments/files/28688650/batmon.logs.bumble.4-4-20.seconds.-.23.38.txt)

After reading that log I have switched to 5-5-20  and will reboot the whole system now !

**Comment by fl4p** [comment]
@typxxi Thanks for the detailed write-up — it pointed straight at two real bugs, both fixed in **1.98d**.

**1. Config silently not saving.** This was my fault: in 1.98b I moved `publish_period` and `expire_values_after` into the schema's optional group, which makes the HA add-on config form tuck them under "Show unused optional configuration options". Edits made to optional/not-yet-set fields in that section get **silently dropped by the HA frontend on save** — a long-standing HA bug ([home-assistant/frontend#8364](https://github.com/home-assistant/frontend/issues/8364)). That's exactly why your 4-4-30 came back as 4-3-20: the visible required field saved, the two collapsed ones reverted. 1.98d puts both back as always-visible required fields, so they persist normally.

**2. The wired/serial path could never open the port** — details in #22.

After updating to 1.98d, please retry and let me know. If a value still won't stick, switch the config editor to YAML mode as a workaround.

**Comment by typxxi** [comment]
good to read that you could follow my observations cause it looked or felt weird watching my "figures" being changed over and over again 

### **I. Another incident of "saving ignores my value" - this time hciX** 

I also recogniced  but missed to note a weird behaviour regarding this hci0 , hci1 ... even hci2 and hci3 variables which also had been ignored during saving - not always but quite often. I typed hci0 and after reboot of the app / add on the hci0 had been ignored and I saw hci1 again. 

And in such cases I could see in the log sometimes that one (hci1 or hci0) was operating in uart and the other in usb mode so I changed that figure to that of the USB related hciX number as I knew that I needed the BT one which should be the USB related hciX path.


### **II. Current summary from today from 06 am till now : looks stablized** 

Somehow tonight from 06 am (I choose 10.000 lines to download and I got over 8500 starting from 06:23) the system seems to have stabilized itself in bluetooth mode with 4 -4 - 20 conifugration.

Meanwhile I had found out that counting "**Power cycle bluetooth hardware**" in the log file (notepad++) shows most likely the amount of "app" reboots which are now down to 0 
from over 13 in less than 1 hour yesterday evening.

In my 9000 line long log from 5 am things look stable, I got a report every 4 seconds which feels normal again.


### **III. From bumble to bluek and 5 - 5 - 20 seconds** 

Reading the log I also recogniced from this line 
`07:17:24 INFO [sampling] battery1 stack: Bleak 3.0.2, bluek/bluez-v5.85`
that I had missed to mention 2 more changes in the early hours of today that might have contributed to that stabilisation:

When I did not saw an improvement from the  **"bumble"** option I went ahead to **"bluek"**
I have also changed the time span to **5 - 5 - 20 seconds combination** 
All of this might have contributed to the improvement, maybe also a new full reboot of the whole HA OS system at 01 am.

### **IV. Hickup in the log ?** 

Here is a kind of hickup I found in that 8500 line log 
```
07:17:05 INFO [sampling] JKBt(C8:47:80:10:08:A1,battery1) times: connect=0.00s fetch=0.43s
07:17:10 INFO [sampling] battery1: BmsSampl(70.0%,U=53.2V,I=13.90A,P=740W,Q=167/333Ah,mos=25°C)
07:17:10 INFO [sampling] battery1 volt=[3326,3324,3325,3324,3325,3325,3324,3326,3324,3325,3325,3326,3326,3326,3324,3325] temp=[23.5, 23.05, 23.6, 23.8]
07:17:10 INFO [sampling] JKBt(C8:47:80:10:08:A1,battery1) times: connect=0.00s fetch=0.46s
07:17:16 INFO [sampling] battery1: BmsSampl(70.0%,U=53.2V,I=13.90A,P=740W,Q=167/333Ah,mos=25°C)
07:17:16 INFO [sampling] battery1 volt=[3327,3326,3326,3325,3326,3326,3326,3326,3325,3325,3325,3327,3326,3326,3325,3326] temp=[23.5, 23.05, 23.65, 23.8]
07:17:16 INFO [sampling] JKBt(C8:47:80:10:08:A1,battery1) times: connect=0.00s fetch=0.44s
07:17:24 WARNING [bt] BMS JKBt(C8:47:80:10:08:A1,battery1) disconnected after 18008.8s!
07:17:24 ERROR [sampling] battery1 error (#1): TimeoutError: timeout waiting for 2 at bmslib/__init__.py:84 in wait_for() <- CancelledError at bmslib/__init__.py:81 in wait_for()
07:17:24 INFO [sampling] battery1 bms debug data: {'resp': {3: (bytearray(b'U\xaa\xeb\x90\x03\xa5JK_PB2A16S20P\x00\x00\x0015A\x00\x00\x00\x00\x0015.38\x00\x00\x00\xf8\xfc\xac\x03\x1e\x00\x00\x00HBB_PB2A16S20\x00\x00\x001234\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00240622\x00\x0040110490690\x00000\x00Input Userdat\x00\x00\x00hannama2\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xfe\xff\xff\xff\x8f\xe9\x1d\x02\x00\x00\x00\x00\x90\x1f\x00\x00\x00\x00\xc0\xd8\xe7\xfe?\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x06\x03\xffg\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\x0f\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\xffg\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x01\x14\x00\x00\x00\x19\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00/\x00\x00\nF\x01\x1e\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xfe\x9f\xe9\xff\x0f\x00\x00\x00\x00\x00\x00\x00G'), 1780877837.1115658), 1: (bytearray(b'U\xaa\xeb\x90\x01\xa0\xd5\x0c\x00\x00\xfb\t\x00\x00T\x0b\x00\x00\x10\x0e\x00\x00\xb1\r\x00\x00\x04\x00\x00\x00\x01\x0e\x00\x00\x0f\n\x00\x00\x06\x0e\x00\x00\x16\r\x00\x00\xf6\t\x00\x00@\r\x03\x00\x03\x00\x00\x00<\x00\x00\x00@\r\x03\x00,\x01\x00\x00<\x00\x00\x00\x05\x00\x00\x00\xd0\x07\x00\x00\xf4\x01\x00\x00\xc2\x01\x00\x00\xf4\x01\x00\x00\xc2\x01\x00\x00\x14\x00\x00\x002\x00\x00\x00\x84\x03\x00\x00\xbc\x02\x00\x00\x10\x00\x00\x00\x01\x00\x00\x00\x01\x00\x00\x00\x01\x00\x00\x00\xc8\x14\x05\x00\xdc\x05\x00\x00(\n\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00`\xe3\x16\x00\x80\x01<2\x18\xfe\xff\xff\xff\x9f\xe9\x1d\x02\x00\x00\x00\x00\xc6'), 1780877836.9636939), 2: (bytearray(b'U\xaa\xeb\x90\x02\x97\xff\x0c\xfd\x0c\xfd\x0c\xfe\x0c\xfe\x0c\xfe\x0c\xfe\x0c\xff\x0c\xfe\x0c\xfd\x0c\xfd\x0c\xff\x0c\xff\x0c\xff\x0c\xfe\x0c\xff\x0c\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff\x00\x00\xfe\x0c\x02\x00\x00\x01=\x00?\x00K\x00L\x00X\x00`\x00n\x00{\x00\x82\x00{\x00r\x00n\x00f\x00O\x00K\x00A\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xfb\x00\x00\x00\x00\x00\xe4\xcf\x00\x00\x82J\x0b\x00P6\x00\x00\xeb\x00\xe6\x00\x00\x00\x00\x00\x00\x00\x00F\xb8\x8d\x02\x00\xb4\xa9\x03\x00\x12\x00\x00\x00v~C\x00H\x00\x00\x00\x8eC\xad\x03\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\x00\x01\x00\x00\x00"\x03(\x00\x00\x00\xd7\xf1@@\x00\x00\x00\x00\xca\x14\x00\x00\x00\x01\x01\x01\x00\x06\x00\x00\x82\x051\x05\x00\x00\x00\x00\xfb\x00\xed\x00\xee\x00e\x03,~\x1a\x0c\xeb\x1d\x00\x00\x80Q\x01\x00\x00\x00\x03\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\xfe\xff\x7f\xdc/\x01\x01\xb0\x0f\x07\x00\x00\xe3'), 1780895838.2435348)}, 'char_w': <bluek.characteristic.BleakGATTCharacteristic object at 0x7f9aed4b00>, 'char_r': <bluek.characteristic.BleakGATTCharacteristic object at 0x7f9aed4b00>}
07:17:24 INFO [sampling] battery1 device info: DeviceInfo(JK_PB2A16S20P,hw-15A,sw-15.38,HBB_PB2A16S20,#40110490690)
07:17:24 INFO [sampling] battery1 stack: Bleak 3.0.2, bluek/bluez-v5.85
07:17:27 INFO [bt] bt_diagnostics C8:47:80:10:08:A1: seen rssi=-52 name='HBB_PB2A16S20-01' adapter=hci0 (5 devices in range)
07:17:27 ERROR [main] 1 exceptions occurred fetching BMSs
07:17:27 ERROR [main] Error (num 1, max 200) reading BMS: TimeoutError: timeout waiting for 2 at bmslib/__init__.py:84 in wait_for() <- CancelledError at bmslib/__init__.py:81 in wait_for()
07:17:33 INFO [bt] connecting battery1 (C8:47:80:10:08:A1) adapter=hci0 timeout=10
07:17:34 INFO [bt] [Service] 1: 00001800-0000-1000-8000-00805f9b34fb
07:17:34 INFO [bt] 	[Characteristic] 2: 00002a00-0000-1000-8000-00805f9b34fb (read,write) (read,write), Value: b''
07:17:34 INFO [bt] 	[Characteristic] 4: 00002a01-0000-1000-8000-00805f9b34fb (read,write) (read,write), Value: b'\x00\x00'
07:17:34 INFO [bt] 	[Characteristic] 6: 00002a02-0000-1000-8000-00805f9b34fb (read) (read), Value: b'\x00'
07:17:34 INFO [bt] 	[Characteristic] 8: 00002a04-0000-1000-8000-00805f9b34fb (read) (read), Value: b'\x0c\x00\x10\x00\x00\x00X\x02'
07:17:34 INFO [bt] [Service] 10: 00001801-0000-1000-8000-00805f9b34fb
07:17:34 INFO [bt] 	[Characteristic] 11: 00002a05-0000-1000-8000-00805f9b34fb (read,indicate) (read,indicate), Value: b'\x01\x00\xff\xff'
07:17:35 INFO [bt] 		[Descriptor] 13: 00002902-0000-1000-8000-00805f9b34fb) | Value: b'\x00\x00'
07:17:35 INFO [bt] [Service] 14: 0000ffe0-0000-1000-8000-00805f9b34fb
07:17:35 INFO [bt] 	[Characteristic] 15: 0000ffe2-0000-1000-8000-00805f9b34fb (write-without-response) (write-without-response), Value: None
07:17:35 INFO [bt] 	[Characteristic] 17: 0000ffe1-0000-1000-8000-00805f9b34fb (write-without-response,write,notify) (write-without-response,write,notify), Value: None
07:17:35 INFO [bt] 		[Descriptor] 19: 00002902-0000-1000-8000-00805f9b34fb) | Value: b'\x00\x00'
07:17:35 INFO [bt] [Service] 20: 0000180a-0000-1000-8000-00805f9b34fb
07:17:35 INFO [bt] 	[Characteristic] 21: 00002a29-0000-1000-8000-00805f9b34fb (read) (read), Value: b'BEKEN SAS'
07:17:35 INFO [bt] 	[Characteristic] 23: 00002a24-0000-1000-8000-00805f9b34fb (read) (read), Value: b'BK-BLE-1.0'
07:17:35 INFO [bt] 	[Characteristic] 25: 00002a25-0000-1000-8000-00805f9b34fb (read) (read), Value: b'1.0.0.0-LE'
07:17:35 INFO [bt] 	[Characteristic] 27: 00002a27-0000-1000-8000-00805f9b34fb (read) (read), Value: b'1.0.0'
07:17:35 INFO [bt] 	[Characteristic] 29: 00002a26-0000-1000-8000-00805f9b34fb (read) (read), Value: b'6.1.2'
07:17:35 INFO [bt] 	[Characteristic] 31: 00002a28-0000-1000-8000-00805f9b34fb (read) (read), Value: b'6.3.0'
07:17:35 INFO [bt] 	[Characteristic] 33: 00002a23-0000-1000-8000-00805f9b34fb (read) (read), Value: b'\x124V\xff\xfe\x9a\xbc\xde'
07:17:35 INFO [bt] 	[Characteristic] 35: 00002a2a-0000-1000-8000-00805f9b34fb (read) (read), Value: b'\xff\xee\xdd\xcc\xbb\xaa'
07:17:35 INFO [bt] 	[Characteristic] 37: 00002a50-0000-1000-8000-00805f9b34fb (read) (read), Value: b'\x02^\x04@\x00\x00\x03'
07:17:35 INFO [bt] [Service] 39: 0000180f-0000-1000-8000-00805f9b34fb
07:17:35 INFO [bt] 	[Characteristic] 40: 00002a19-0000-1000-8000-00805f9b34fb (read,notify) (read,notify), Value: b'\x00'
07:17:35 INFO [bt] 		[Descriptor] 42: 00002902-0000-1000-8000-00805f9b34fb) | Value: b'\x00\x00'
07:17:35 INFO [bt] [Service] 43: f000ffc0-0451-4000-b000-000000000000
07:17:35 INFO [bt] 	[Characteristic] 44: f000ffc1-0451-4000-b000-000000000000 (write-without-response,write,notify) (write-without-response,write,notify), Value: None
07:17:35 INFO [bt] 		[Descriptor] 46: 00002902-0000-1000-8000-00805f9b34fb) | Value: b'\x00\x00'
07:17:35 INFO [bt] 		[Descriptor] 47: 00002901-0000-1000-8000-00805f9b34fb) | Value: b'Img Identify\x00'
07:17:35 INFO [bt] 	[Characteristic] 48: f000ffc2-0451-4000-b000-000000000000 (write-without-response,write,notify) (write-without-response,write,notify), Value: None
07:17:35 INFO [bt] 		[Descriptor] 50: 00002902-0000-1000-8000-00805f9b34fb) | Value: b'\x00\x00'
07:17:35 INFO [bt] 		[Descriptor] 51: 00002901-0000-1000-8000-00805f9b34fb) | Value: b'Img Block\x00'
07:17:37 INFO [sampling] connected bms JKBt(C8:47:80:10:08:A1,battery1)!
07:17:37 INFO [sampling] battery1: BmsSampl(70.0%,U=53.2V,I=14.43A,P=768W,Q=167/333Ah,mos=25°C)
07:17:37 INFO [sampling] battery1 volt=[3330,3329,3329,3329,3330,3330,3329,3330,3329,3329,3327,3330,3330,3330,3329,3330] temp=[23.5, 23.05, 23.65, 23.8]
07:17:37 INFO [sampling] JKBt(C8:47:80:10:08:A1,battery1) times: connect=3.55s fetch=0.56s
07:17:43 INFO [sampling] battery1: BmsSampl(70.0%,U=53.3V,I=14.95A,P=796W,Q=167/333Ah,mos=25°C)
07:17:43 INFO [sampling] battery1 volt=[3331,3330,3330,3329,3330,3330,3330,3330,3330,3329,3329,3331,3331,3330,3329,3330] temp=[23.5, 23.05, 23.65, 23.8]
```
But it looks like the add on and bluetooth stack could stabilize itself again.

I will now look for the latest add on upgrade but also stop writing here and report about the UART and batmon story here [#22](https://github.com/fl4p/batmon-ha/issues/22).
